R/databricks.R

```diff
@@ -0,0 +1,159 @@
+    # this field.
+    user_agent <- Sys.getenv("SPARK_CONNECT_USER_AGENT")
+  }
+
+  args <- list(
+    odbc(),
+    driver = driver,
+    host = hostname,
+    HTTPPath = http_path,
+    ThriftTransport = 2,
+    UserAgentEntry = user_agent,
+    # Connections to Databricks are always over HTTPS.
+    port = 443,
+    Protocol = "https",
+    SSL = 1
+  )
+
+  # Check for Workbench-provided credentials.
+  if (exists(".rs.api.getDatabricksToken")) {
+    wb_token <- .rs.api.getDatabricksToken(workspace)
+  }
+
+  if (nchar(token) != 0) {
+    # An explicit PAT takes precedence over everything else.
+    args <- c(args, AuthMech = 3, uid = "token", pwd = token)
+  } else if (nchar(client_id) != 0) {
+    # Next up are explicit OAuth2 M2M credentials.
+    args <- c(
+      args, AuthMech = 11, Auth_Flow = 1, Auth_Client_ID = client_id,
+      Auth_Client_Secret = client_secret
+    )
+  } else if (length(wb_token) != 0 && nchar(wb_token) != 0) {
+    # Next up are Workbench-provided credentials.
+    args <- c(args, AuthMech = 11, Auth_Flow = 0, Auth_AccessToken = wb_token)
+  } else if (!is_hosted_session() && nchar(Sys.which(cli_path)) != 0) {
+    # When on desktop, try using the Databricks CLI for auth.
+    output <- suppressWarnings(
+      system2(
+        cli_path, c("auth", "token", "--host", workspace), stdout = TRUE,
+        stderr = TRUE
+      )
+    )
+    output <- paste(output, collapse = "\n")
+    # If we don't get an error message, try to extract the token from the JSON-
+    # formatted output.
+    if (grepl("access_token", output, fixed = TRUE)) {
+      token <- gsub(".*access_token\":\\s?\"([^\"]+).*", "\\1", output)
+      args <- c(args, AuthMech = 11, Auth_Flow = 0, Auth_AccessToken = token)
+    }
+  }
+
+  args
+}
+
+# Try to determine whether we can redirect the user's browser to a server on
+# localhost, which isn't possible if we are running on a hosted platform.
+#
+# This is based on the strategy pioneered by the {gargle} package and {httr2}.
+is_hosted_session <- function() {
+  # If RStudio Server or Posit Workbench is running locally (which is possible,
+  # though unusual), it's not acting as a hosted environment.
+  Sys.getenv("RSTUDIO_PROGRAM_MODE") == "server" &&
+    !grepl("localhost", Sys.getenv("RSTUDIO_HTTP_REFERER"), fixed = TRUE)
+}
+
+# Returns a sensible driver name even if odbc.ini and odbcinst.ini do not
+# contain an entry for the Databricks ODBC driver.
+default_databricks_driver <- function() {
+  # For Linux and macOS we can default to the shared library paths used by the
+  # official installers. On Windows we use the official driver name instead.
+  default_path <- ""
+  if (Sys.info()["sysname"] == "Linux") {
+    default_path <- "/opt/simba/spark/lib/64/libsparkodbc_sb64.so"
+  } else if (Sys.info()["sysname"] == "Darwin") {
+    default_path <- "/Library/simba/spark/lib/libsparkodbc_sbu.dylib"
+  }
+
+  if (file.exists(default_path)) default_path else "Simba Spark ODBC Driver"
+}
```

Comment:
Yay for sensible defaults.

For linux, would we want to default to `/opt/rstudio-drivers/spark/bin/lib/libsparkodbc_sb64.so` and fallback to the `/opt/simba/spark` location?

Reply:
Is that the expected install location for the Pro driver? That could make sense.

Reply:
Yes, that is the expected location of the driver. I don't quite understand `bin/lib` in the path, but it follows the convention of where the other drivers are installed.